MMI Instruction Implementation

astrelsky opened the pull request:
I have added in the instructions in my repo to this one. The cspec still is not quite right. "BADSPACEBASE" still appears in some functions when the sp is being stored/retrieved. It comes and goes when changing the stackpointer register to/from sp_lo/sp.

Commit:
Added proper implementation of MMI instructions.

Fixed and squashed 10/19/2019

Ignore VSCode files

Added special case for dsra32

Added proper implementation of MMI instructions.

Minor changes were also made to the registers as necessary for
the MMI implementations.

beardypig commented:
I'll need some time to look at this properly. My only concern is that the special cases of `PADDUB`, etc. were done to improve decompilation - if they operate on the 128 bit registers you loose that improvement and get poor decompilation.

astrelsky commented:
The special cases are still there. The varnode size limitation doesn't affect a direct = so I changed the special cases of moveq to be RD128 = RS128src;

Take all the time you need, it's a lot.

beardypig commented:
It's not that there is a limit to the size of a varnode, it's that there is a limit to the size of a constant varnode. Which does cause some ugly decompilation, when I have a moment I'll check the specifics.

astrelsky commented:
> It's not that there is a limit to the size of a varnode, it's that there is a limit to the size of a constant varnode. Which does cause some ugly decompilation, when I have a moment I'll check the specifics.

I see. Will definitely have to add as many special cases as possible. It doesn't come out too bad. If you run a search for memchr you will quickly come across some mmi instructions. In all honesty, considering what the instructions actually do, it's the unnecessary typecasting that is the most ugly.

beardypig commented on data/languages/r5900.cspec:
```diff
@@ -18,86 +18,82 @@
           <entry size="2" alignment="2" />
           <entry size="4" alignment="4" />
-          <entry size="8" alignment="4" />
+          <entry size="8" alignment="8" />
```
Do they definitely have to be 8 byte aligned?

astrelsky replied:
100% positive that longs are 8 bytes and 8 byte aligned.

beardypig commented on build.gradle:
```diff
@@ -31,3 +31,13 @@ else {
 }
 //----------------------END "DO NOT MODIFY" SECTION-------------------------------
+
+def sleigh = new File(ghidraInstallDir).getCanonicalPath() + "/support/sleigh.bat"
```
Unfortunately this won't work on an OS other than Windows... To compile the sleigh files you use `ant` in the `data` directory (which build.xml is).

Commit:
Corrected/Added missing syscalls

GregoryMorse committed:
IDA Pro binding added

beardypig committed:
Use gradlew to install the latest version of gradle

beardypig committed:
Update readme to include some installation instructions

beardypig commented on build.gradle:
```diff
@@ -31,3 +31,21 @@ else {
+
+def sleigh = new File(ghidraInstallDir).getCanonicalPath() + sleigh_path
+task runSleigh(type:Exec) {
+    commandLine sleigh, "-a", "data/languages"
```
Is compiling the `sla` files required? Doesn't Ghidra do it when you load the extension?

astrelsky replied:
It does now. I don't think ghidra did initially. Until any java code is added build.gradle shouldn't be necessary.

beardypig replied:
Ah, perhaps. I will check on an older version. The gradle file is used to package the extension in to a zip too.

beardypig commented:
Do you think this is good to merge?

astrelsky commented:
> Do you think this is good to merge?

Should be good to go now.

astrelsky commented:
I realized I misunderstood the bitrange assignments. Don't merge this yet I have to go through and fix it. I thought it was like a slice but it is really register[{start_bit}, {range_length}]
Also they are bits, not bytes. I have work to do.

astrelsky commented:
Ok, I have rewritten them and simplified them as best as possible. For library functions that were written in assembly with the MMI instructions you may see a lot of "CONCAT" but it is just the nature of what is going on and the varnode size restriction.

lq and sq produce a warning about the inaccurate size, but they decompile correctly otherwise.

beardypig commented:
@astrelsky this needs a rebase too :(

Commit:
Merge branch 'master' of https://github.com/astrelsky/ghidra-emotionengine

# Conflicts:
#	data/languages/r5900.cspec

astrelsky commented:
I handled the conflicts as well as did some updates. It is still possible to encounter CONCAT at times. When this is encountered it is usually because it was a hand-written asm routine instead of compiler optimized.

beardypig commented:
Cool. I don't think the `CONCAT`s are much of a problem :)
